components/style/gecko/data.rs

```diff
@@ -113,30 +109,29 @@ impl PerDocumentStyleData {
 }
 
 impl PerDocumentStyleDataImpl {
-    /// Recreate the style data if the stylesheets have changed.
-    pub fn flush_stylesheets(&mut self) {
-        let mut stylist = if self.device_changed || self.stylesheets_changed {
-            Some(Arc::get_mut(&mut self.stylist).unwrap())
-        } else {
-            None
-        };
-
-        if self.device_changed {
-            Arc::get_mut(&mut stylist.as_mut().unwrap().device).unwrap().reset();
-            self.device_changed = false;
-            // Force a stylesheet flush if the device has changed.
-            self.stylesheets_changed = true;
+    /// Reset the device state because it may have changed.
+    ///
+    /// Implies also a stylesheet flush.
+    pub fn reset_device(&mut self) {
+        {
+            let mut stylist = Arc::get_mut(&mut self.stylist).unwrap();
+            Arc::get_mut(&mut stylist.device).unwrap().reset();
         }
+        self.stylesheets_changed = true;
```

Comment:
I didn't want to do the restyle stuff after I've finished all the medialist work, because gecko has its own control about which media queries have matched and not, and can provide us with an adequate restyle hint if we do the right thing.

Servo does indeed have the concept of "setting a new device", but there are fundamental differences between a servo device and a Gecko device that don't allow us to do that for gecko.

Concretely, in gecko right now the device is represented by the pres context and effectively unique. There's no easy way to compare against a previous state of the pres context (and I think its not bad, but we can't just reuse that piece of code so easily)